src/main/java/com/flab/offcoupon/service/coupon_use/OrderService.java

```diff
@@ -0,0 +1,132 @@
+package com.flab.offcoupon.service.coupon_use;
+
+import com.flab.offcoupon.domain.entity.OrderDetail;
+import com.flab.offcoupon.domain.entity.params.AppliedCouponInfo;
+import com.flab.offcoupon.domain.entity.params.OrderInfo;
+import com.flab.offcoupon.domain.vo.persistence.order.AvailableCouponsByMemberIdVo;
+import com.flab.offcoupon.domain.vo.persistence.order.CouponIssuesAreActiveVo;
+import com.flab.offcoupon.domain.vo.persistence.order.MemberIdProductIdNowVo;
+import com.flab.offcoupon.domain.vo.persistence.order.ValidateNowIsBetweenPeriodVo;
+import com.flab.offcoupon.dto.request.OrderProductRequest;
+import com.flab.offcoupon.dto.response.AvailableCouponsByMemberIdResponse;
+import com.flab.offcoupon.exception.coupon.CouponStatusException;
+import com.flab.offcoupon.exception.coupon.CouponUsageInvalidPeriodException;
+import com.flab.offcoupon.repository.mysql.*;
+import com.flab.offcoupon.util.ResponseDTO;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.flab.offcoupon.domain.entity.OrderCoupon.createOrderCoupon;
+import static com.flab.offcoupon.domain.entity.OrderDetail.createOrderDetail;
+import static com.flab.offcoupon.domain.entity.params.OrderInfo.createOrderInfo;
+import static com.flab.offcoupon.exception.coupon.CouponErrorMessage.COUPON_IS_NOT_ACTIVE;
+import static com.flab.offcoupon.exception.coupon.CouponErrorMessage.COUPON_USAGE_INVALID_PERIOD;
+
+@RequiredArgsConstructor
+@Service
+public class OrderService {
+
+    private final CouponIssueRepository couponIssueRepository;
+    private final CouponRepository couponRepository;
+    private final ProductRepository productRepository;
+    private final OrderDetailRepository orderDetailRepository;
+    private final OrderCouponRepository orderCouponRepository;
+
+    /**
+     * 사용 가능한 쿠폰 목록 조회
+     * @param memberId 회원 ID
+     * @param productId 상품 ID
+     * @param now 현재 날짜
+     * @return 사용 가능한 쿠폰 목록
+     */
+    @Transactional(readOnly = true)
+    public ResponseDTO<List<AvailableCouponsByMemberIdResponse>> getAvailableCoupons(final long memberId, final long productId, final LocalDateTime now) {
+        List<AvailableCouponsByMemberIdVo> availableCoupons =
+                couponIssueRepository.getAvailableCoupons(new MemberIdProductIdNowVo(memberId, productId, now));
+        List<AvailableCouponsByMemberIdResponse> responseList = new ArrayList<>();
+        filterAvailableCouponsWithMinPrice(productId, availableCoupons, responseList);
```

Comment:
Collection을 생성한 후 메서드 내부에서 컬랙션의 상태를 변경시키고 있네요.
이렇게 내부에서 상태를 변경시키는 메서드는 코드를 읽을 때 해당 변수에 대해서 계속해서 트래킹하게 만들기 때문에 피로감을 느끼게 만들고, 메서드의 결과값을 예측할 수 없게 만듭니다.

그렇기 때문에 상태를 변경시키는 것이 아닌, 행위 그 자체를 메서드화한다면 좀 더 깔끔하고 읽기 쉬운 직관적인 코드로 만들 수 있습니다.

```suggestion
        List<AvailableCouponsByMemberIdResponse> responseList = availableCoupons.stream
        .filter(coupon -> isOverThenMinPrice(productId))
        ...
```

Reply:
말씀해주신 filter를 사용해보려고 했는데, 할인 가격이 minOrderPrice를 초과하지 않는 범위까지만 누적하는거라 filter 사용이 어려울것 같습니답...